trie/fog.py

```diff
@@ -0,0 +1,284 @@
+            else:
+                return nearest_right
+
+    def nearest_right(self, key_input: NibblesInput) -> Nibbles:
+        """
+        Find the foggy prefix that is nearest on the right to the supplied key.
+
+        :raises IndexError: if there are no foggy prefixes to the right
+        """
+        if isinstance(key_input, Nibbles):
+            key = key_input
+        else:
+            key = Nibbles(key_input)
+
+        index = self._unexplored_prefixes.bisect(key)
+        if index == 0:
+            return self._unexplored_prefixes[0]
+        else:
+            nearest_left = self._unexplored_prefixes[index - 1]
+
+            # always return nearest right, unless prefix of key is unexplored
+            if key_starts_with(key, nearest_left):
+                return nearest_left
+            else:
+                # This can raise a IndexError if index == len(unexplored prefixes)
+                # Raising a IndexError is the desired behavior in this case!
+                return self._unexplored_prefixes[index]
+
+    @staticmethod
+    @to_tuple
+    def _prefix_distance(low_key: Nibbles, high_key: Nibbles) -> Tuple[int, ...]:
+        """
+        How far are the two keys from each other, as a sequence of differences.
+        The first non-zero distance must be positive, but the remaining distances may
+        be negative. Distances are designed to be simply compared, like distance1 < distance2.
+
+        The high_key must be higher than the low key, or the output distances are not
+        guaranteed to be accurate.
+        """
+        for low_nibble, high_nibble in zip_longest(low_key, high_key, fillvalue=None):
+            if low_nibble is None:
+                final_low_nibble = 15
+            else:
+                final_low_nibble = low_nibble
+
+            if high_nibble is None:
+                final_high_nibble = 0
+            else:
+                final_high_nibble = high_nibble
+
+            # Note: this might return a negative value. It's fine, because only the
+            #   relative distance matters. For example (1, 2) and (2, 1) produce a
+            #   distance of (1, -1). If the other reference point is (3, 1), making
+            #   the distance to the middle (1, 0), then the "correct" thing happened.
+            #   The (1, 2) key is a tiny bit closer to the (2, 1) key, and a tuple
+            #   comparison of the distance will show it as a smaller distance.
+            yield final_high_nibble - final_low_nibble
+
+    @classmethod
+    def _new_trie_fog(cls, unexplored_prefixes: SortedSet) -> 'HexaryTrieFog':
+        """
+        Convert a set of unexplored prefixes to a proper HexaryTrieFog object.
+        """
+        copy = cls()
+        copy._unexplored_prefixes = unexplored_prefixes
+        return copy
+
+    def serialize(self) -> bytes:
+        prefixes = list(self._unexplored_prefixes)
+        return f"HexaryTrieFog:{prefixes!r}".encode()
```

Comment:
Missed this on my first pass too.

Any reason you chose to do the encoding this way rather than an actual binary encoding of the prefixes?  Seems like it would take up significantly less space if they were all packed tightly.

Reply:
Nope, this was just fast. :+1: for compressing them